### PR TITLE
filter bad resolvers

### PR DESCRIPTION
**Pull request by ohnorobo**
Roughly filter DNS resolvers that aren't reliable because they have too many errors.

Dashboards: [1](https://datastudio.google.com/c/u/0/reporting/bcba5de4-0402-4687-8b7c-1ad1b04ad3d1/page/p_0foeibtt0c), [2](https://datastudio.google.com/c/reporting/fa204158-932b-43bf-b3e8-b346c1468792/page/p_mekso35lrc)

This is better than the [previous filtering solution](https://github.com/censoredplanet/censoredplanet-analysis/commit/4a28e74bf2aa13cdacba48523de7500c46207661) which was just a hacky query.

### CODE REVIEW

File: table/queries/derived_satellite_scans.sql [hunk 1]
```diff
@@ -73,6 +73,20 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
   ), ",")
 );
 
+CREATE TEMP FUNCTION BadResolver(
```
Comment (fortuna) [inline table/queries/derived_satellite_scans.sql:76]: Could you explain (perhaps in a comment), what those rates are? What are the numerator and denominator?

We need our evaluation to be understandable easily, so we need to explain these rates.

File: table/queries/derived_satellite_scans.sql [hunk 2]
```diff
@@ -73,6 +73,20 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
...
+  resolver_private_ip_rate FLOAT64,
+  resolver_zero_ip_rate FLOAT64,
+  resolver_connect_error_rate FLOAT64,
+  resolver_invalid_cert_rate FLOAT64
+) AS (
+  (resolver_non_zero_rcode_rate +
+   resolver_private_ip_rate +
+   resolver_zero_ip_rate +
+   resolver_connect_error_rate +
+   resolver_invalid_cert_rate) > 0.15
```
Comment (fortuna) [inline table/queries/derived_satellite_scans.sql:87]: It looks like the chart you used to derive this number is broken. Please revise.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: table/queries/derived_satellite_scans.sql (hunk 1 above)
Comment (ohnorobo) [reply table/queries/derived_satellite_scans.sql:80]: Added a comment

File: table/queries/derived_satellite_scans.sql [hunk 3]
```diff
@@ -73,6 +73,24 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
...
+# number_of_errors_on_resolver/total_measurements_on_resolver
+# resolvers which consistantly return too many errors are filtered out.
+CREATE TEMP FUNCTION BadResolver(
+  resolver_non_zero_rcode_rate FLOAT64,
+  resolver_private_ip_rate FLOAT64,
+  resolver_zero_ip_rate FLOAT64,
+  resolver_connect_error_rate FLOAT64,
+  resolver_invalid_cert_rate FLOAT64
+) AS (
+  (resolver_non_zero_rcode_rate +
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply table/queries/derived_satellite_scans.sql:87]: I would rather leave it in for now and then set the rate correctly once it's fixed.

File: table/queries/derived_satellite_scans.sql [hunk 4]
```diff
@@ -73,6 +73,24 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
   ), ",")
 );
 
+# resolver_*_rate fields measure various error rates for each resolver across the whole scan
+# They should be a value from 0-1 calculated as
+# number_of_errors_on_resolver/total_measurements_on_resolver
+# resolvers which consistantly return too many errors are filtered out.
+CREATE TEMP FUNCTION BadResolver(
+  resolver_non_zero_rcode_rate FLOAT64,
```
[the remark this answers is not in the document]
Comment (ohnorobo) [reply table/queries/derived_satellite_scans.sql:81]: Not all of the resolvers problems are just instability. Some of them server very high rates of captive portals, or other errors

**Comment by ohnorobo** [comment]
Switched to >.75 for three of the 5 rates only per offline discussion.

File: table/queries/derived_satellite_scans.sql [hunk 5]
```diff
@@ -73,6 +73,21 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
...
+# They should be a value from 0-1 calculated as
+# number_of_unique_domains_with_error_on_resolver_per_scan
+#  /total_number_of_unique_domains_on_resolver_per_scan
+# resolvers which consistantly return too many errors are filtered out.
+CREATE TEMP FUNCTION BadResolver(
+  resolver_connect_error_rate FLOAT64,
+  resolver_invalid_cert_rate FLOAT64,
+  resolver_non_zero_rcode_rate FLOAT64
+) AS (
+  resolver_connect_error_rate > .75 OR
```
Comment (fortuna) [inline table/queries/derived_satellite_scans.sql:86]: Could you add a comment explaining that this is timeout?

File: table/queries/derived_satellite_scans.sql [hunk 6]
```diff
@@ -73,6 +73,21 @@ CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
...
+  resolver_invalid_cert_rate > .75 OR
```
Comment (fortuna) [inline table/queries/derived_satellite_scans.sql:87]: I think this also deserves a comment, explaining it's from one of the fetches?

**Review by fortuna — APPROVED** [review]

File: table/queries/derived_satellite_scans.sql (hunk 5 above)
Comment (ohnorobo) [reply table/queries/derived_satellite_scans.sql:88]: done

File: table/queries/derived_satellite_scans.sql (hunk 6 above)
Comment (ohnorobo) [reply table/queries/derived_satellite_scans.sql:89]: done

**Pull request merged**